File: loadercomplex-forge/src/ver122/java/com/enderzombi102/loadercomplex/forge12/impl/ForgeAddonContext.java
package com.enderzombi102.loadercomplex.forge12.impl;

import com.enderzombi102.loadercomplex.api.AddonContext;
import com.enderzombi102.loadercomplex.api.platform.*;
import com.enderzombi102.loadercomplex.forge12.impl.utils.ForgeFactoryWorld;
import com.enderzombi102.loadercomplex.forge12.impl.utils.ForgeI18nSystem;
import com.enderzombi102.loadercomplex.forge12.impl.utils.ForgeRegistry;
import net.minecraft.launchwrapper.Launch;
import net.minecraftforge.common.ForgeVersion;
import net.minecraftforge.fml.common.FMLCommonHandler;
import net.minecraftforge.fml.common.Loader;
import org.jetbrains.annotations.NotNull;

import java.nio.file.Path;

public class ForgeAddonContext implements AddonContext {
	private final ForgeRegistry registry;
	private final I18nSystem i18nSystem;

	public ForgeAddonContext() {
		this.registry = new ForgeRegistry();
		this.i18nSystem = new ForgeI18nSystem();
	}

	@Override
	public @NotNull PlatformInfo getPlatformInfo() {
		return new PlatformInfo() {
			@Override
			public @NotNull String name() {
				return "Forge";
			}

			@Override
			public @NotNull String version() {
				return ForgeVersion.getVersion();
			}

			@Override
			public @NotNull String id() {
				return "forge-legacy";
			}

			@Override
			public @NotNull String minecraftVersion() {
				return "1.12.2";
			}

			@Override
			public @NotNull String getApiVersion() {
				return "0.1.3";
			}

			@Override
			public boolean isDeveloperEnvironment() {
				return (boolean) Launch.blackboard.getOrDefault( "fml.deobfuscatedEnvironment", false );
			}
		};
	}

	@Override
	public @NotNull Registry getRegistry() {
		return this.registry;
	}

	@Override
	public @NotNull FactoryWorld getFactoryWorld() {
		return new ForgeFactoryWorld();
	}

	@Override
	public boolean isAtLeastMinecraft( String version ) {
		throw new RuntimeException( "Not implemented" );
	}

	@Override
	public boolean isDedicatedServer() {
		return FMLCommonHandler.instance().getSide().isServer();
	}

	@Override
	public boolean isModLoaded( String id ) {
		return Loader.isModLoaded( id );
	}

	@Override
	public @NotNull I18nSystem getI18nSystem() {
		return this.i18nSystem;
	}

	@Override
	public @NotNull ResourceLoader getResourceLoader() {
		throw new RuntimeException( "Not implemented" );
	}

	@Override
	public @NotNull Path getConfigDir() {
		return Loader.instance().getConfigDir().toPath();
	}
}

File: loadercomplex-forge/src/ver122/kotlin/com/enderzombi102/loadercomplex/forge122/impl/ForgeAddonContext.kt
package com.enderzombi102.loadercomplex.forge122.impl

import com.enderzombi102.loadercomplex.api.AddonContext
import com.enderzombi102.loadercomplex.api.minecraft.command.CommandManager
import com.enderzombi102.loadercomplex.api.minecraft.keybind.KeybindManager
import com.enderzombi102.loadercomplex.api.minecraft.network.NetworkManager
import com.enderzombi102.loadercomplex.api.platform.*
import com.enderzombi102.loadercomplex.forge122.impl.platform.ForgeFactoryWorld
import com.enderzombi102.loadercomplex.forge122.impl.platform.ForgeI18nSystem
import com.enderzombi102.loadercomplex.forge122.impl.platform.ForgeRegistry
import com.unascribed.flexver.FlexVerComparator
import net.minecraft.launchwrapper.Launch
import net.minecraftforge.common.ForgeVersion
import net.minecraftforge.fml.common.FMLCommonHandler
import net.minecraftforge.fml.common.Loader
import java.nio.file.Path

class ForgeAddonContext : AddonContext {
	private val registry = ForgeRegistry()
	private val i18nSystem = ForgeI18nSystem()

	override fun getPlatformInfo(): PlatformInfo {
		return object : PlatformInfo {
			override fun name(): String =
				"Forge"

			override fun version(): String =
				ForgeVersion.getVersion()

			override fun id(): String =
				"forge-legacy"

			override fun minecraftVersion(): String =
				"1.12.2"

			override fun getApiVersion(): String =
				"0.1.3"

			override fun isDeveloperEnvironment(): Boolean =
				Launch.blackboard.getOrDefault("fml.deobfuscatedEnvironment", false) as Boolean
		}
	}

	override fun getRegistry(): Registry =
		this.registry

	override fun getFactoryWorld(): FactoryWorld =
		ForgeFactoryWorld()

	override fun isAtLeastMinecraft(version: String): Boolean =
		FlexVerComparator.compare( ForgeVersion.mcVersion, version ) <= 0

	override fun isDedicatedServer(): Boolean =
		FMLCommonHandler.instance().side.isServer

	override fun isModLoaded(id: String): Boolean =
		Loader.isModLoaded(id)

	override fun getI18nSystem(): I18nSystem =
		this.i18nSystem

	override fun getResourceLoader(): ResourceLoader =
		throw NotImplementedError()

	override fun getConfigDir(): Path =
		Loader.instance().configDir.toPath()

	override fun getKeybindManager(): KeybindManager =
		throw NotImplementedError()

	override fun getNetworkManager(): NetworkManager =
		throw NotImplementedError()

	override fun getCommandManager(): CommandManager =
		throw NotImplementedError()
}
